см readme
C:\Users\специалист ао\mypandas\work\2023\23-09-ХХ Переоценка ДК магазины выст\ReadMe2023-10-13.docx

Код 

- подтягивания поля вставки из файла 01-01-2023 стандартную выгрузку из 1С



# imports

In [412]:
import numpy as np
import pandas as pd
import re
import os

In [413]:
from rusprice.dictdiamonds import dcol_dia, dcla_dia, dpo_dia, dogr_dia_discount
#import rusprice.func_rpdiamonds_xlsx_v2022_2_CARtoKVG_v2 as func_rpdiamonds_v2
#func_rpdiamonds_xlsx_v2023_1_CARtoKVG.py
import rusprice.func_rpdiamonds_xlsx_v2023_1_CARtoKVG as func_rpdiamonds_v2

In [414]:
folder = '../'
#cm readme

folder_res = folder+'for_script/'
folder_res

'../for_script/'

# Входные данные

## из 1С

In [415]:
#1. Вводные/

In [416]:
os.listdir(folder+'1 Вводные/')

['выгрузка УТ 2023-08-10',
 'Выгрузка УТ 2023-09-19',
 'выгрузка УТ 2023-10-12',
 'выгрузка УТ 2023-10-15',
 'Выгрузки ЮТД 2023-10-13']

In [417]:
files = os.listdir(folder+'1 Вводные/выгрузка УТ 2023-10-15')
files

['Остатки1СУТ__все_склады_ДК_2023-10-15_закупочные.xlsx',
 'Остатки1СУТ__все_склады_ДК_2023-10-15_розница.xlsx']

In [418]:
list(enumerate(files))

[(0, 'Остатки1СУТ__все_склады_ДК_2023-10-15_закупочные.xlsx'),
 (1, 'Остатки1СУТ__все_склады_ДК_2023-10-15_розница.xlsx')]

In [419]:
filename_z = files[0]
filename_r= files[1]
filename_z, filename_r

('Остатки1СУТ__все_склады_ДК_2023-10-15_закупочные.xlsx',
 'Остатки1СУТ__все_склады_ДК_2023-10-15_розница.xlsx')

In [426]:
df_z = pd.read_excel(folder+'1 Вводные/выгрузка УТ 2023-10-15/'+filename_z,header=9)#, dtype={5: str}

In [427]:
#Вывести итог таблицы
df_z.iloc[len(df_z)-1][[ 'Количество конечный остаток', 'Конечный остаток вес',
       'Оценка (конечный остаток)']]

Количество конечный остаток        34,269.00
Конечный остаток вес              122,668.35
Оценка (конечный остаток)     880,921,254.30
Name: 34286, dtype: object

на 09-15-2023 ут

Количество конечный остаток        34,269.00
Конечный остаток вес              122,668.35
Оценка (конечный остаток)     880,921,254.30
Name: 34286, dtype: object
на 09-10-2023 ут
Количество конечный остаток         34334.0
Конечный остаток вес              122885.73
Оценка (конечный остаток)      883196152.97
Name: 34350, dtype: object

на 19-09-2023 ут?
Количество конечный остаток         34528.0
Конечный остаток вес              123441.14
Оценка (конечный остаток)      888714246.96
Name: 34544, dtype: object

In [428]:
df_z.dropna(subset=['Код'], inplace=True)

In [429]:
#Вывести сумм таблицы
df_z[[ 'Количество конечный остаток', 'Конечный остаток вес',
       'Оценка (конечный остаток)']].sum()

Количество конечный остаток        34,269.00
Конечный остаток вес              122,668.35
Оценка (конечный остаток)     880,921,254.30
dtype: float64

на 13-10-2023 ут?
Количество конечный остаток    3.433400e+04
Конечный остаток вес           1.228857e+05
Оценка (конечный остаток)      8.831962e+08
dtype: float64
на 19-09-2023 ут?
Количество конечный остаток    3.452800e+04
Конечный остаток вес           1.234411e+05
Оценка (конечный остаток)      8.887142e+08
dtype: float64


In [430]:
# Загрузка файла с розничными ценами

In [431]:
df_r = pd.read_excel(folder+'1 Вводные/выгрузка УТ 2023-10-15/'+filename_r,header=9)#, dtype={5: str}

In [349]:
#Вывести итог таблицы
df_r.iloc[len(df_r)-1][[ 'Количество конечный остаток', 'Конечный остаток вес',
       'Оценка (конечный остаток)']]

Количество конечный остаток          34311.0
Конечный остаток вес               122805.47
Оценка (конечный остаток)      2913591188.89
Name: 34327, dtype: object

In [350]:
на 13-10-2023
Количество конечный остаток          34334.0
Конечный остаток вес               122885.73
Оценка (конечный остаток)      2915863314.89
Name: 34350, dtype: object

SyntaxError: invalid syntax (<ipython-input-350-aac4f57e1a38>, line 1)

In [ ]:
на 19-09-2023 ут?
Количество конечный остаток          34528.0
Конечный остаток вес               123441.14
Оценка (конечный остаток)      2935241509.89
Name: 34544, dtype: object

In [351]:
-34334.0+34528.0, 2935241509.89-2915863314.89, 8.887142e+08-8.831962e+08

(194.0, 19378195.0, 5518000.0)

In [352]:
df_r.dropna(subset=['Код'], inplace=True)

In [353]:
#Вывести сумм таблицы
df_r[[ 'Количество конечный остаток', 'Конечный остаток вес',
       'Оценка (конечный остаток)']].sum()

Количество конечный остаток    3.431100e+04
Конечный остаток вес           1.228055e+05
Оценка (конечный остаток)      2.913591e+09
dtype: float64

In [354]:
Количество конечный остаток    3.433400e+04
Конечный остаток вес           1.228857e+05
Оценка (конечный остаток)      2.915863e+09
dtype: float64

SyntaxError: invalid syntax (<ipython-input-354-46a72082c1e3>, line 1)

In [ ]:
на 19-09-2023 ут?
Количество конечный остаток    3.452800e+04
Конечный остаток вес           1.234411e+05
Оценка (конечный остаток)      2.935242e+09
dtype: float64

In [355]:
df_r.rename(columns={'Оценка (конечный остаток)':'розничная стоимость'}, inplace=True)

In [356]:
df_z.rename(columns={'Оценка (конечный остаток)':'закупочная стоимость'}, inplace=True)

In [357]:
#проверка соответствия кодов
len(df_r['Код'].unique()) ,len(df_r)

(34311, 34311)

In [358]:
len(df_z['Код'].unique()) ,len(df_z)

(34311, 34311)

In [359]:
set(df_z['Код'].unique()) == set(df_r['Код'].unique())

True

In [360]:
#Присоединяем к стоимости df_1c
df_1c = pd.merge(df_r,df_z[['Код', 'закупочная стоимость']]
                           ,how='left',right_on='Код',left_on='Код')

In [366]:
pd.options.display.float_format = '{:,.2f}'.format

In [367]:
df_1c[['закупочная стоимость','розничная стоимость']].sum()

закупочная стоимость     882,526,213.42
розничная стоимость    2,913,591,188.89
dtype: float64

In [368]:
df_1c.columns

Index(['Склад', 'Unnamed: 1', 'Unnamed: 2', 'Код', 'Unnamed: 4',
       'Старый код (Справочник "Номенклатура" (Общие))', 'Склад.1',
       'Склад.Склад / Магазин', 'Unnamed: 8',
       'Вставки (Справочник "Номенклатура" (Общие))', 'Вид номенклатуры',
       'Тип номенклатуры', 'Номенклатура', 'Группа',
       'Заводской номер (Справочник "Номенклатура" (Общие))',
       'Проба (Справочник "Номенклатура" (Общие))',
       'Цвет металла (Справочник "Номенклатура" (Общие))',
       'Дизайн (Справочник "Номенклатура" (Общие))', 'Артикул ',
       'Размерная группа Центрального камня (Справочник "Номенклатура" (Общие))',
       'Цвет Центрального Камня (Справочник "Номенклатура" (Общие))',
       'Чистота Центрального Камня (Справочник "Номенклатура" (Общие))',
       'Размер (Справочник "Номенклатура" (Общие))', 'Наименование для печати',
       'Единица измерения веса',
       'Вес Вставок ПДК (Карат) (Справочник "Номенклатура" (Общие))',
       'Кол-во ДК (Справочник "Номенклатура" (Об

In [369]:
for x in ['Склад', 'Unnamed: 1', 'Unnamed: 2', 'Unnamed: 4','Unnamed: 8','Склад.1','Unnamed: 33',  'Unnamed: 29', 'Unnamed: 30', 'Unnamed: 31','Unnamed: 32']:
    print(x, df_1c[x].unique()[:13])

Склад [nan]
Unnamed: 1 [nan]
Unnamed: 2 [nan]
Unnamed: 4 [nan]
Unnamed: 8 [nan]
Склад.1 ['Благовещенск' 'Выставочный' 'Интернет-магазин' 'Иркутск' 'Кружало'
 'Ленина, 28' 'Ленина23' 'Ленск' 'Мирный' 'Московское представительство'
 'Новосибирск' 'Пассаж' 'Хабаровск']
Unnamed: 33 [nan 'Сахаювелир ООО']
Unnamed: 29 ['Ювелирные изделия' nan 'Драгоценности Якутии ООО' 'МИЗО'
 'Сахаювелир ООО' 'Кристалл 2001  ООО' 'СУВЕНИРЫ' 'Золото Якутии ОАО АПЮК'
 'Разные поставщики' '0']
Unnamed: 30 ['Саха Минералс Лимитед' 'Ювелирные изделия' 'ИП "Алексеев В.М."'
 '№ 06-329/2 от 12.12.16' 'Драгоценности Якутии ООО' 'Сахаювелир ООО'
 '№03/10-212 от 05.11.07' '№6/2-08-КП от 14.01.2008г.' 'Невский Бриллиант'
 'Золото Якутии ОАО АПЮК' 'Сапфир М ООО' '№11/08-КП от 21.01.2008г.'
 'Покровский гранильный завод ООО']
Unnamed: 31 ['Sakha Minerals Limited SM21-01 от 28.09.2021' 'Из Москвы'
 '10-271 от 21.12.2018' '06-329/2 от 12.12.2016 Задание №4'
 '06-341/4 от 26.12.2016' '03/10-95 от 15.05.08' '89/1/06 от 20.04

In [370]:
df_1c.drop(['Склад', 'Unnamed: 1', 'Unnamed: 2','Unnamed: 4','Unnamed: 8'], axis=1, inplace=True)

In [371]:
df_1c.rename(columns={'Склад.1':'Склад','Unnamed: 29':'родитель1_1c', 'Unnamed: 30':'родитель2_1c',
                      'Unnamed: 31':'родитель3_1c', 'Unnamed: 32':'родитель4_1c', 'Unnamed: 33':'родитель5_1c'},
            inplace=True)

In [372]:
df_1c

Код Старый код (Справочник "Номенклатура" (Общие))  \
0      ЦБ-00200503                                            NaN   
1      ЦБ-00002854                                     МС00057368   
2      ЦБ-00019951                                     МС00152100   
3      ЦБ-00019953                                     МС00146311   
4      ЦБ-00111074                                     МС00111422   
...            ...                                            ...   
34306  ЦБ-00182016                                     0000032928   
34307  ЦБ-00182017                                     0000032929   
34308  ЦБ-00182019                                     0000032934   
34309  ЦБ-00182020                                     0000032935   
34310  ЦБ-00182021                                     0000032936   

                   Склад Склад.Склад / Магазин  \
0           Благовещенск     Розничный магазин   
1           Благовещенск     Розничный магазин   
2           Благовещенск     Розничный магазин   
3           Благовещенск     Розничный магазин   
4           Благовещенск     Розничный магазин   
...                  ...                   ...   
34306  Ювелирные изделия         Оптовый склад   
34307  Ювелирные изделия         Оптовый склад   
34308  Ювелирные изделия         Оптовый склад   
34309  Ювелирные изделия         Оптовый склад   
34310  Ювелирные изделия         Оптовый склад   

             Вставки (Справочник "Номенклатура" (Общие)) Вид номенклатуры  \
0                                      45КР57-0,520-4/5А     Т ДМ Юи с ДК   
1      2Кр57-0,083-2/2А 4Кр57-0,084-2/2А 6Кр57-0,280-...     Т ДМ Юи с ДК   
2                                       1Кр57-0,086-4/5А     П ДМ Юи с ДК   
3                                        3Кр57-0,06-4/5А     П ДМ Юи с ДК   
4      1Кр57-0,050-4/5А 1Кр57-0,031-6/6А 1Кр57-0,014-...     Т ДМ Юи с ДК   
...                                                  ...              ...   
34306                                                NaN     Т ДМ Юи с ДК   
34307                                                NaN     Т ДМ Юи с ДК   
34308                                                NaN     Т ДМ Юи с ДК   
34309                                                NaN     Т ДМ Юи с ДК   
34310                                                NaN     Т ДМ Юи с ДК   

      Тип номенклатуры Номенклатура                           Группа  \
0                Товар      Браслет  спецификация №1 от 27.01.2022г.   
1                Товар      Браслет                         Товар ДК   
2                Товар        Брошь                            Броши   
3                Товар        Брошь                          накл.№5   
4                Товар        Брошь                            51 ДК   
...                ...          ...                              ...   
34306            Товар  серьги с ДК        Акт №1 АК "Золото Якутии"   
34307            Товар  серьги с ДК        Акт №1 АК "Золото Якутии"   
34308            Товар  серьги с ДК        Акт №1 АК "Золото Якутии"   
34309            Товар  серьги с ДК        Акт №1 АК "Золото Якутии"   
34310            Товар  серьги с ДК        Акт №1 АК "Золото Якутии"   

      Заводской номер (Справочник "Номенклатура" (Общие))  ...  \
0                                                    512   ...   
1                                               30575,51   ...   
2                                                1191890   ...   
3                                                   1094   ...   
4                                                    422   ...   
...                                                  ...   ...   
34306                                           01-12308   ...   
34307                                           01-12309   ...   
34308                                           01-14045   ...   
34309                                           01-14046   ...   
34310                                           01-14047   ...   

  

In [373]:
#Если без самородков, пропуск 3
df_1c_res = df_1c

# Подтянуть массу самородков

## подтянуть масса самородков из '3.1. Самородки Сводка.xlsx',

In [33]:
files = os.listdir(folder+'1. Вводные/Веса самородков/')
files

FileNotFoundError: [WinError 3] Системе не удается найти указанный путь: '//UR-JVV/Users/DAS/Desktop/GM TMP/Analytics/Проекты в работе/23-09-ХХ Переоценка ДК магазины выст/1. Вводные/Веса самородков/'

In [ ]:
[x for x in enumerate(files)]

In [ ]:
df_sm1 = pd.read_excel(folder+'1. Вводные/Веса самородков/'+files[1])#,header=9)#, dtype={5: str}

In [ ]:
df_sm1.dropna(subset=['№ столбца'], inplace=True)

In [ ]:
 df_sm1[['Код', 'Артикул', 'Вес изделия','Вес самородка (общий суммированный)']]

In [ ]:
#проверка кода
len(df_sm1['Код'].unique()), len(df_sm1)

In [ ]:
#Присоединяем к df_1c
df_1c = pd.merge(df_1c ,df_sm1[['Код', 'Вес самородка (общий суммированный)']]
                           ,how='left',right_on='Код',left_on='Код')

In [ ]:
#присоеденил массу самородка к df_1c, проверяем сколько не получило массу 
df_1c['Вес самородка (общий суммированный)'].isna().sum()

In [ ]:
#Кто без цены
df_1c[df_1c['Вес самородка (общий суммированный)'].isna()]

In [ ]:
#df_1c['Проба самородка'].unique() -> array([894.,  nan, 885., 932., 889., 810., 890., 820., 875.])

## подтянуть масса самородков из '22-09-23 ДК и самородки На утверждение.xlsm'

### считываем файл срезаем самородки

In [ ]:
files = os.listdir(folder+'1. Вводные/Веса самородков/')
files

In [ ]:
[x for x in enumerate(files)]

In [ ]:
df_sm2 = pd.read_excel(folder+'1. Вводные/Веса самородков/'+files[0])#,header=9)#, dtype={5: str}

In [ ]:
df_sm2.columns

In [ ]:
df_sm2[['Наименование изделия', 'Артикул', 'Проба','Вес, грамм', 'Учётная стоимость','Вес золотого самородка (общий суммированный)']]

In [ ]:
df_sm2.dropna(subset=['Вес золотого самородка (общий суммированный)'], inplace=True)

### Создаем ключи и присоединяем массу самородка

In [ ]:
#проблема в том что данный файл без ключа, поэтому подтягиваем по совокупному совпадению - 
#- 'Номенклатура','Артикул ', 'Конечный остаток вес'

In [ ]:
#проверка уникальности ключа
len((df_sm2['Наименование изделия']+df_sm2['Артикул'].astype('str')+df_sm2['Вес, грамм'].astype('str')).unique())

In [ ]:
len(df_sm2)#.dropna(subset=['Вес, грамм']))

In [ ]:
df_1c_33 = df_1c[df_1c['Вес самородка (общий суммированный)'].isna()].copy()

In [ ]:
df_1c_33['key'] = df_1c_33[['Номенклатура','Артикул ',
  'Конечный остаток вес']].apply(lambda x: x['Номенклатура']+str(x['Артикул '])+str(x['Конечный остаток вес']),
                                 axis=1)

In [ ]:
df_sm2['key'] = df_sm2[['Наименование изделия', 'Артикул',
        'Вес, грамм']].apply(lambda x: x['Наименование изделия']+str(x['Артикул'])+str(x['Вес, грамм']),
                                 axis=1)

In [ ]:
len(set(df_sm2['key'])),len(df_sm2['key'])

In [ ]:
len(set(df_1c_33['key'])),len(df_1c_33['key'])

In [ ]:
#Этих из наличия нет в 27 утвержденных 9 шт из них  "Самородок Севера" понятно, тк это имитация  
set(df_1c_33['key']) - set(df_sm2['key'])

In [ ]:
#Эти есть в утверждении но нет в наличии - видимо проданы 
set(df_sm2['key'])-set(df_1c_33['key'])

In [ ]:
#set(df_1c_33['key']).intersection(set(df_sm2['key']))

In [ ]:
#Присоединяем к df_1c
df_1c_33 = pd.merge(df_1c_33,
         df_sm2[['Вес золотого самородка (общий суммированный)','key']],
         how='left',
         right_on=['key'],
         left_on=['key']
         )

In [ ]:
#Список без массы самородка
df_1c_33[df_1c_33['Вес золотого самородка (общий суммированный)'].isna()]

In [ ]:
df_1c_454 = df_1c[~df_1c['Вес самородка (общий суммированный)'].isna()].copy()

In [ ]:
folder_res

In [ ]:
#запись промрезультата для проверки на excel
pd.concat([df_1c_454,df_1c_33]).to_excel(folder_res+'df_1c_ms.xlsx')

## Сборка файла результата сборки весов самородков

In [ ]:
df_1c_454.columns

In [ ]:
df_1c_33.columns

In [ ]:
df_1c_33['вес самородка'] = df_1c_33['Вес золотого самородка (общий суммированный)']
df_1c_454['вес самородка'] = df_1c_454['Вес самородка (общий суммированный)']

In [ ]:
df_1c_res = pd.concat([df_1c_454,df_1c_33])

In [ ]:
df_1c_res['вес самородка'].isna().sum()

In [ ]:
#Вывести сумм таблицы
df_1c_res[[ 'Количество конечный остаток', 'Конечный остаток вес',
           'розничная стоимость', 'закупочная стоимость', 'вес самородка'
          ]].sum()

должно быть
Количество конечный остаток         487.00
Конечный остаток вес               3537.87
Оценка (конечный остаток)      12375195.96
dtype: float64

In [ ]:
df_1c_res['источник веса самородка'] = np.nan

In [ ]:
df_1c_res.loc[(~df_1c_res['Вес золотого самородка (общий суммированный)'].isna()),
              'источник веса самородка'] =  '22-09-23 ДК и самородки На утверждение.xlsm'

In [ ]:
df_1c_res.loc[(~df_1c_res['Вес самородка (общий суммированный)'].isna()),'источник веса самородка'] = '3.1. Самородки Сводка.xlsx'

In [ ]:
df_1c_res.columns

In [ ]:
df_1c_res = df_1c_res[['Код', 'Старый код (Справочник "Номенклатура" (Общие))', 'Склад',
       'Склад.Склад / Магазин', 'Вставки (Справочник "Номенклатура" (Общие))',
       'Вид номенклатуры', 'Тип номенклатуры', 'Номенклатура', 'Группа',
       'Заводской номер (Справочник "Номенклатура" (Общие))',
       'Проба (Справочник "Номенклатура" (Общие))',
       'Цвет металла (Справочник "Номенклатура" (Общие))',
       'Дизайн (Справочник "Номенклатура" (Общие))', 'Артикул ',
       'Размерная группа Центрального камня (Справочник "Номенклатура" (Общие))',
       'Цвет Центрального Камня (Справочник "Номенклатура" (Общие))',
       'Чистота Центрального Камня (Справочник "Номенклатура" (Общие))',
       'Размер (Справочник "Номенклатура" (Общие))', 'Наименование для печати',
       'Единица измерения веса',
       'Вес Вставок ПДК (Карат) (Справочник "Номенклатура" (Общие))',
       'Кол-во ДК (Справочник "Номенклатура" (Общие))',
       'Вес Вставок ДК (Карат) (Справочник "Номенклатура" (Общие))', 'Группы',
       'родитель1', 'родитель2', 'родитель3', 'родитель4',
       'Количество конечный остаток', 'Конечный остаток вес',
       'розничная стоимость', 'закупочная стоимость',
       'вес самородка',
       'источник веса самородка']]

# Подтянуть вставки  ШК, родитель 1 -5 из 2023-01-01 Назовем их Вставки010123

o	Данные по вставкам надежнее чем с 1с (на 1 янв 2023г) \\UR-JVV\Users\DAS\Desktop\GM TMP\Analytics\Проекты в работе\23-09-XX Переоценка Самородки\1. Вводные\Данные по вставкам\23-01-01 Сводка ЮИ.xlsx

In [374]:
files = os.listdir('//UR-JVV/Users/DAS/Desktop/GM TMP/Analytics/Проекты в работе/23-09-XX Переоценка Самородки/1. Вводные/Данные по вставкам/')
files

['2.4 Сводка Описания вставок V2.xlsx', '23-01-01 Сводка ЮИ.xlsx']

In [375]:
list(enumerate(files))

[(0, '2.4 Сводка Описания вставок V2.xlsx'), (1, '23-01-01 Сводка ЮИ.xlsx')]

In [376]:
fn = files[1]
fn
# 1, '23-01-01 Сводка ЮИ.xlsx' основной файл на 01-01-2023, 
# 0  '2.4 Сводка Описания вставок V2.xlsx' - неисправный файл был отклонен

'23-01-01 Сводка ЮИ.xlsx'

In [377]:
df_vst = pd.read_excel('//UR-JVV/Users/DAS/Desktop/GM TMP/Analytics/Проекты в работе/23-09-XX Переоценка Самородки/1. Вводные/Данные по вставкам/'\
                       +fn)#,header=9)#, dtype={5: str}

In [378]:
df_vst.columns

Index(['Source.Name', 'Штрихкод', 'Старый код', 'Код', 'Наименование',
       'Артикул', 'Зав Ном', 'Вид ном', 'Проба', 'Цвет металла', 'Кол', 'Вес',
       'Закуп', 'Розн', 'Вставки', 'Размер', 'РазмернаяГруппаЦК', 'ЦветЦК',
       'ЧистотаЦК', 'Дизайн', 'Родитель1', 'Родитель2', 'Родитель3',
       'Родитель4', 'Родитель5'],
      dtype='object')

In [379]:
len(df_vst['Код'].unique()), len(df_vst)

(143861, 143861)

In [380]:
len(df_1c_res['Код'].unique()), len(df_1c_res)

(34311, 34311)

In [381]:
#Присоединяем к df_1c
df_1c_res = pd.merge(df_1c_res,
         df_vst[['Код','Вставки', 'Штрихкод', 'Старый код',
                 'Родитель1', 'Родитель2', 'Родитель3',
                 'Родитель4', 'Родитель5' ]],
         how='left',
         right_on=['Код'],
         left_on=['Код']
         )

In [382]:
#223 пустых строк это по 3 договорам новые Валив и ДЯ
df_1c_res['Штрихкод'].isna().sum()

222

In [383]:
df_1c_res.rename(columns={'Вставки':'Вставки010123'}, inplace=True)

# Запись результата в файл (не используется) 

In [93]:
folder_res

'//UR-JVV/Users/DAS/Desktop/GM TMP/Analytics/Проекты в работе/23-09-ХХ Переоценка ДК магазины выст/for_script/'

In [44]:
#запись промрезультата для проверки на excel
#df_1c_res.to_excel(folder_res+'Остатки1С__все_склады_ДК_2023-09-19.xlsx')

# Анализ 0

In [384]:
df_1c_res.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 34311 entries, 0 to 34310
Data columns (total 41 columns):
 #   Column                                                                   Non-Null Count  Dtype  
---  ------                                                                   --------------  -----  
 0   Код                                                                      34311 non-null  object 
 1   Старый код (Справочник "Номенклатура" (Общие))                           28274 non-null  object 
 2   Склад                                                                    34311 non-null  object 
 3   Склад.Склад / Магазин                                                    34311 non-null  object 
 4   Вставки (Справочник "Номенклатура" (Общие))                              24326 non-null  object 
 5   Вид номенклатуры                                                         34311 non-null  object 
 6   Тип номенклатуры                                                      

In [385]:
df_1c_res.columns

Index(['Код', 'Старый код (Справочник "Номенклатура" (Общие))', 'Склад',
       'Склад.Склад / Магазин', 'Вставки (Справочник "Номенклатура" (Общие))',
       'Вид номенклатуры', 'Тип номенклатуры', 'Номенклатура', 'Группа',
       'Заводской номер (Справочник "Номенклатура" (Общие))',
       'Проба (Справочник "Номенклатура" (Общие))',
       'Цвет металла (Справочник "Номенклатура" (Общие))',
       'Дизайн (Справочник "Номенклатура" (Общие))', 'Артикул ',
       'Размерная группа Центрального камня (Справочник "Номенклатура" (Общие))',
       'Цвет Центрального Камня (Справочник "Номенклатура" (Общие))',
       'Чистота Центрального Камня (Справочник "Номенклатура" (Общие))',
       'Размер (Справочник "Номенклатура" (Общие))', 'Наименование для печати',
       'Единица измерения веса',
       'Вес Вставок ПДК (Карат) (Справочник "Номенклатура" (Общие))',
       'Кол-во ДК (Справочник "Номенклатура" (Общие))',
       'Вес Вставок ДК (Карат) (Справочник "Номенклатура" (Общие))', 'Гру

In [386]:
for c in ['Родитель1', 'Родитель2', 'Родитель3', 'Родитель4', 'Родитель5']:
    print(c, '-->', df_1c_res.sample(10)[c])

Родитель1 --> 2821     Ювелирные изделия
27194    Ювелирные изделия
2505     Ювелирные изделия
3980     Ювелирные изделия
3731     Ювелирные изделия
11859    Ювелирные изделия
33531    Ювелирные изделия
28191    Ювелирные изделия
551      Ювелирные изделия
2835     Ювелирные изделия
Name: Родитель1, dtype: object
Родитель2 --> 7675     Драгоценности Якутии ООО
16644    Драгоценности Якутии ООО
9759        Саха Минералс Лимитед
31178                  ООО "РИОН"
31576    Драгоценности Якутии ООО
10916       Саха Минералс Лимитед
25548      Золото Якутии ОАО АПЮК
16489                        МИЗО
29202    Драгоценности Якутии ООО
29247          ИП "Алексеев В.М."
Name: Родитель2, dtype: object
Родитель3 --> 18241                       №11/08-КП от 21.01.2008г.
9196                         SM19-01 от 27.09.2019 ДК
1311                         №10-048 от 05.04.2022 г.
2669                             SM10/01 от 07.10.10г
6305                      SM10-03 от 14.07.2011 доп.1
27296           

In [286]:
for c in ['родитель1_1c', 'родитель2_1c', 'родитель3_1c', 'родитель4_1c', 'родитель5_1c']:
    print(c, '-->\n', df_1c_res.sample(10)[c])

родитель1_1c -->
 1607            Ювелирные изделия
2177                          NaN
31717          Кристалл 2001  ООО
16667           Ювелирные изделия
30711           Ювелирные изделия
2683            Ювелирные изделия
32815           Ювелирные изделия
29035           Ювелирные изделия
9648     Драгоценности Якутии ООО
22137                         NaN
Name: родитель1_1c, dtype: object
родитель2_1c -->
 3256                          Ювелирные изделия
18129    Договор комиссии №03/10-31 от 28.02.02
29691                №6/2-08-КП от 14.01.2008г.
4772                         ИП "Алексеев В.М."
14838                         Ювелирные изделия
22285                         Ювелирные изделия
7677                         ИП "Алексеев В.М."
20308                 №11/08-КП от 21.01.2008г.
2792                         ИП "Алексеев В.М."
31505    Договор комиссии №03/10-31 от 28.02.02
Name: родитель2_1c, dtype: object
родитель3_1c -->
 20082                           2 партия
31345            

In [387]:
df_1c_res.value_counts('Склад')

Склад
Ювелирные изделия               16506
Центральный                      1773
Иркутск                          1764
Ленина23                         1726
Пассаж                           1638
Кружало                          1595
Ленина, 28                       1560
Новосибирск                      1404
Хабаровск                        1336
Ленск                            1247
Мирный                           1140
Благовещенск                     1036
Интернет-магазин                  861
Выставочный                       675
Московское представительство       50
dtype: int64

на 19-09-2023 ут?
Склад
Ювелирные изделия               16505
Центральный                      1793
Иркутск                          1772
Ленина23                         1750
Пассаж                           1619
Ленина, 28                       1584
Кружало                          1564
Новосибирск                      1413
Хабаровск                        1343
Ленск                            1253
Мирный                           1156
Благовещенск                     1046
Интернет-магазин                  861
Выставочный                       818
Московское представительство       51
dtype: int64

In [388]:
#Где находятся 223 с не подключенным штрихкодом
df_1c_res[df_1c_res['Штрихкод'].isna()].value_counts('Склад')

Склад
Ювелирные изделия               171
Московское представительство     42
Кружало                           3
Центральный                       3
Ленина, 28                        2
Пассаж                            1
dtype: int64

In [389]:
#Где находятся 214 с не подключенным Сводка Вставки
df_1c_res[df_1c_res['Вставки010123'].isna()].value_counts('Склад')

Склад
Ювелирные изделия               171
Московское представительство     32
Кружало                           3
Центральный                       3
Ленина, 28                        2
Пассаж                            1
dtype: int64

In [390]:
#Где находятся 213 с не подключенным полем вставки
df_1c_res[df_1c_res['Вставки010123'].isna()].value_counts('Склад')

Склад
Ювелирные изделия               171
Московское представительство     32
Кружало                           3
Центральный                       3
Ленина, 28                        2
Пассаж                            1
dtype: int64

In [391]:
#По каким договорам 223 с не подключенным полем Вставки
df_1c_res[df_1c_res['Вставки010123'].isna()].value_counts(['родитель2_1c','родитель3_1c','родитель4_1c'])

родитель2_1c              родитель3_1c                родитель4_1c       
Драгоценности Якутии ООО  №10-171 от 13.10.2022       РТ-5 от 24.10.2022     113
                          №103/03-2022 от 27.12.2022  Рт-25 от 28.12.2022     66
ОСОО "Валив"              №V22-01 от 07.07.2022       эксклюзив               33
dtype: int64

In [392]:
#По каким договорам 223 с не подключенным полем штрихкод
df_1c_res[df_1c_res['Штрихкод'].isna()].value_counts(['родитель2_1c','родитель3_1c','родитель4_1c'])

родитель2_1c              родитель3_1c                родитель4_1c        
Драгоценности Якутии ООО  №10-171 от 13.10.2022       РТ-5 от 24.10.2022      113
                          №103/03-2022 от 27.12.2022  Рт-25 от 28.12.2022      66
ОСОО "Валив"              №V22-01 от 07.07.2022       эксклюзив                33
Ювелирные изделия         ИП "Алексеев В.М."          10-064 от 07.04.2021     10
dtype: int64

In [393]:
df_1c_res[df_1c_res['Вставки010123'].isna()].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 212 entries, 5325 to 34241
Data columns (total 41 columns):
 #   Column                                                                   Non-Null Count  Dtype  
---  ------                                                                   --------------  -----  
 0   Код                                                                      212 non-null    object 
 1   Старый код (Справочник "Номенклатура" (Общие))                           0 non-null      object 
 2   Склад                                                                    212 non-null    object 
 3   Склад.Склад / Магазин                                                    212 non-null    object 
 4   Вставки (Справочник "Номенклатура" (Общие))                              212 non-null    object 
 5   Вид номенклатуры                                                         212 non-null    object 
 6   Тип номенклатуры                                                     

In [296]:
#df_1c_res[df_1c_res['Сводка Вставки'].isna()]

In [394]:
#Проверка на срезанность вставок по 100 символов
df_1c_res.loc[df_1c_res['Вставки010123'].isna(),'Вставки (Справочник "Номенклатура" (Общие))'].map(len).unique()

array([ 16,  33,  35, 103, 127, 105,  37,  34,  87,  70,  52,  73,  51,
        69, 107, 122,  85, 104,  71, 120, 137,  36,  55,  17,  31, 119,
        24], dtype=int64)

# Подтянуть недостающие вставки из вставок 1С

In [395]:
df_1c_res.loc[df_1c_res['Вставки010123'].isna(),'Вставки010123'] = df_1c_res['Вставки (Справочник "Номенклатура" (Общие))']

In [396]:
df_1c_res.loc[df_1c_res['Вставки010123'].isna(),'Вставки010123']

Series([], Name: Вставки010123, dtype: object)

In [397]:
df_1c_res[['Вставки010123']].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 34311 entries, 0 to 34310
Data columns (total 1 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Вставки010123  34311 non-null  object
dtypes: object(1)
memory usage: 1.5+ MB


# Подтянуть недостающие Поставщик, договор, накладная из 1C

In [398]:
list_rod_rod_1c = [('Родитель2','родитель2_1c'),
 ('Родитель3','родитель3_1c'),
 ('Родитель4','родитель4_1c'),
 ('Родитель5','родитель5_1c')]
 

In [302]:
for rod,rod_1c in list_rod_rod_1c:
   df_1c_res.loc[df_1c_res[rod].isna(),rod] = df_1c_res.loc[df_1c_res[rod].isna(),rod_1c]

In [399]:
df_1c_res.loc[df_1c_res['Родитель2'].isna(),'Родитель2'] = df_1c_res.loc[df_1c_res['Родитель2'].isna(),'родитель2_1c']

In [400]:
df_1c_res.loc[df_1c_res['Родитель2'].isna(),'Родитель2']

Series([], Name: Родитель2, dtype: object)

In [401]:
df_1c_res[['Родитель2']].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 34311 entries, 0 to 34310
Data columns (total 1 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Родитель2  34311 non-null  object
dtypes: object(1)
memory usage: 1.5+ MB


## проверка правильности присоеденения недостающих Поставщик, договор, накладная из 1C

родитель2_1c              родитель3_1c                родитель4_1c        
Драгоценности Якутии ООО  №10-171 от 13.10.2022       РТ-5 от 24.10.2022      113
                          №103/03-2022 от 27.12.2022  Рт-25 от 28.12.2022      66
ОСОО "Валив"              №V22-01 от 07.07.2022       эксклюзив                33
Ювелирные изделия         ИП "Алексеев В.М."          10-064 от 07.04.2021     10
dtype: int64

In [402]:
df_1c_res.loc[df_1c_res['Родитель3'] == '10-064 от 07.04.2021']

Код Старый код (Справочник "Номенклатура" (Общие))  \
11604  ЦБ-00096408                                            NaN   
11605  ЦБ-00096400                                            NaN   
11606  ЦБ-00096401                                            NaN   
11607  ЦБ-00096402                                            NaN   
11608  ЦБ-00096403                                            NaN   
11609  ЦБ-00096404                                            NaN   
11610  ЦБ-00096406                                            NaN   
11611  ЦБ-00096407                                            NaN   
11612  ЦБ-00096409                                            NaN   
11653  ЦБ-00096405                                            NaN   

                              Склад Склад.Склад / Магазин  \
11604  Московское представительство         Оптовый склад   
11605  Московское представительство         Оптовый склад   
11606  Московское представительство         Оптовый склад   
11607  Московское представительство         Оптовый склад   
11608  Московское представительство         Оптовый склад   
11609  Московское представительство         Оптовый склад   
11610  Московское представительство         Оптовый склад   
11611  Московское представительство         Оптовый склад   
11612  Московское представительство         Оптовый склад   
11653  Московское представительство         Оптовый склад   

             Вставки (Справочник "Номенклатура" (Общие)) Вид номенклатуры  \
11604  1Груша-5,130-8-5/7аА 3КР57-0,100-3/5А 1КР57-0,...     П ДМ Юи с ДК   
11605  1Радиант-5,130-6/8А 1КР57-0,220-3/5А 1КР57-0,2...     П ДМ Юи с ДК   
11606  1Овал-9,040-8-4/7А 10КР57-1,110-4/5А 1КР57-0,1...     П ДМ Юи с ДК   
11607               1Кушон-5,130-8-5/6Б 6КР57-0,320-2/2А     П ДМ Юи с ДК   
11608                1Кушон-5,010-1/6А 60КР57-0,940-4/3А     П ДМ Юи с ДК   
11609  1КР57-10,090-7/6А 11КР17-0,070-2/4А 4КР17-0,03...     П ДМ Юи с ДК   
11610               1КР57-8,010-8-4/5А 20КР57-1,170-3/2А     П ДМ Юи с ДК   
11611  1Принцесса-5,010-6/7аА 4КР57-0,060-4/3А 1КР57-...     П ДМ Юи с ДК   
11612  1КР57-8,000-8-4/7А 21КР17-0,170-3/4А 7КР17-0,0...     П ДМ Юи с ДК   
11653  1КР57-5,590-8-1/7А 36КР57-1,950-2/2А 26КР57-1,...     П ДМ Юи с ДК   

      Тип номенклатуры Номенклатура                Группа  \
11604            Товар        Колье  10-064 от 07.04.2021   
11605            Товар       Кольцо  10-064 от 07.04.2021   
11606            Товар       Кольцо  10-064 от 07.04.2021   
11607            Товар       Кольцо  10-064 от 07.04.2021   
11608            Товар       Кольцо  10-064 от 07.04.2021   
11609            Товар       Кольцо  10-064 от 07.04.2021   
11610            Товар       Кольцо  10-064 от 07.04.2021   
11611            Товар       Кольцо  10-064 от 07.04.2021   
11612            Товар       Кольцо  10-064 от 07.04.2021   
11653            Товар       Серьги  10-064 от 07.04.2021   

      Заводской номер (Справочник "Номенклатура" (Общие))  ...  \
11604                                            1201378   ...   
11605                                            1201369   ...   
11606                                            1201371   ...   
11607                                            1201372   ...   
11608                                            1201373   ...   
11609                                            1201374   ...   
11610                                            1201376   ...   
11611                                            1201377   ...   
11612                                            1201379   ...   
11653                                            1201375   ...   

       розничная стоимость закупочная стоимость  \
11604        25,804,328.00           545,341.98   
11605        12,884,141.00           166,057.47   
11606        26,728,069.00           594,560.89   
11607        11,068,247.00           255,955.52   
11608        37,085,777.00         1,290,177.12   
11609       135,311,565.00    

In [403]:
df_1c_res['Родитель3'].unique()

array(['Sakha Minerals Limited SM21-01 от 28.09.2021', 'Товар ДК',
       '10-271 от 21.12.2018', '№ 06-329/2 от 12.12.16',
       '06-341/4 от 26.12.2016', '03/10-95 от 15.05.08',
       '89/1/06 от 20.04.06', '№03/10-212 от 05.11.07',
       '№6/2-08-КП от 14.01.2008г.', '06-329/2 от 12.12.16 задание 6',
       'SM10-04 от 11.10.2011 доп2', '№10-159 от 30.06.14',
       '06-13 от 24.01.2011г.', 'SM10-03 от 14.07.2011 доп.1',
       '№03/10-214 от 01.09.06', '№03/10-151 от 30.06.06 г.',
       '№10-121 от 29.03.12 Доп.10 от 29.03.13',
       'SM12-01 от 30.03.12 доп2',
       '№10-206 от 07.07.2016 доп.1 от 30.07.16',
       '№10-141 от 30.03.12 доп.7 от 04.03.13  дав.ДК',
       '10-184 от 06.10.17', '10-246 от 20.12.2019',
       '10-012 от 10.07.2019',
       '№10-141 от 30.03.12 доп.10 от 01.08.13  дав.',
       'SM10/01 от 07.10.10г', 'Договор №03/10-246 от 18.10.06',
       '№11/08-КП от 21.01.2008г.',
       '№10-135 от 30.03.2012г. (давальческий)', '10-100 от 24.05.2019',
    

# Файл для вставок из ЮТД Вставкиютд

## подготовка

In [410]:
files = os.listdir('../1 Вводные/Выгрузки ЮТД 2023-10-13/')
list(enumerate(files))

[(0, '2023-10-13_Остатки товаров на складах (без шк).xlsx'),
 (1, 'Штрихкод - НоменклатураКод (ЦБ) 2023-10-11.xlsx')]

In [411]:
df_utd = pd.read_excel('../1 Вводные/Выгрузки ЮТД 2023-10-13/'+files[0])#,header=9)#, dtype={5: str}

In [406]:
df_utd.iloc[3:]

Unnamed: 0  Unnamed: 1 Unnamed: 2     Unnamed: 3  \
3       Номенклатура.Код         NaN        NaN          Склад   
4                    NaN         NaN        NaN            NaN   
5              000005200         NaN        NaN   Благовещенск   
6              000005201         NaN        NaN   Благовещенск   
7              000005202         NaN        NaN   Благовещенск   
...                  ...         ...        ...            ...   
148823         000186413         NaN        NaN  УР Бриллианты   
148824         000186414         NaN        NaN  УР Бриллианты   
148825         000186415         NaN        NaN  УР Бриллианты   
148826         000186416         NaN        NaN  УР Бриллианты   
148827             Итого         NaN        NaN            NaN   

                    Unnamed: 4              Unnamed: 5  \
3       Серия номенклатуры.УИН  Серия номенклатуры.Вес   
4                          NaN                     NaN   
5             6432205037320158                   47.57   
6             6432205037445837                   15.94   
7             6432205037404896                    4.10   
...                        ...                     ...   
148823        6432205148666206                    1.15   
148824        6432205150175616                    2.22   
148825        6432205150096331                    1.02   
148826        6432205148853228                    4.10   
148827                     NaN                     NaN   

                            Unnamed: 6  Unnamed: 7  \
3       Серия номенклатуры.Вес металла         NaN   
4                                  NaN         NaN   
5                                47.57         NaN   
6                                15.94         NaN   
7                                 4.10         NaN   
...                                ...         ...   
148823                            1.15         NaN   
148824                            2.22         NaN   
148825                            1.02         NaN   
148826                            4.10         NaN   
148827                             NaN         NaN   

                                      Unnamed: 8  \
3       Серия номенклатуры.Вес металла в чистоте   
4                                            NaN   
5                                          44.01   
6                                          14.75   
7                                           3.79   
...                                          ...   
148823                                       NaN   
148824                                       NaN   
148825                                       NaN   
148826                                       NaN   
148827                                       NaN   

                                               Unnamed: 9  ...  \
3       Серия номенклатуры.Вес металла в чистоте ГИИСДМДК  ...   
4                                                     NaN  ...   
5                                                   44.01  ...   
6                                                   14.75  ...   
7                                                    3.79  ...   
...                                                   ...  ...   
148823                                                NaN  ...   
148824                                                NaN  ...   
148825                                                NaN  ...   
148826                                                NaN  ...   
148827                                                NaN  ...   

                                 Unnamed: 14            Unnamed: 15  \
3       Серия номенклатуры.Владелец.Артикул   Номенклатура.Артикул    
4                                        NaN                    NaN   
5                                НЦ 22-002-6            НЦ 22-002-6   
6                                НЦ 22-203-6            НЦ 22-203-6   
7                                НЦ 22-078-6            НЦ 22-078-6   
...                            

In [407]:
df_utd = df_utd.iloc[3:].copy(deep=True)

In [408]:
df_utd

Unnamed: 0  Unnamed: 1 Unnamed: 2     Unnamed: 3  \
3       Номенклатура.Код         NaN        NaN          Склад   
4                    NaN         NaN        NaN            NaN   
5              000005200         NaN        NaN   Благовещенск   
6              000005201         NaN        NaN   Благовещенск   
7              000005202         NaN        NaN   Благовещенск   
...                  ...         ...        ...            ...   
148823         000186413         NaN        NaN  УР Бриллианты   
148824         000186414         NaN        NaN  УР Бриллианты   
148825         000186415         NaN        NaN  УР Бриллианты   
148826         000186416         NaN        NaN  УР Бриллианты   
148827             Итого         NaN        NaN            NaN   

                    Unnamed: 4              Unnamed: 5  \
3       Серия номенклатуры.УИН  Серия номенклатуры.Вес   
4                          NaN                     NaN   
5             6432205037320158                   47.57   
6             6432205037445837                   15.94   
7             6432205037404896                    4.10   
...                        ...                     ...   
148823        6432205148666206                    1.15   
148824        6432205150175616                    2.22   
148825        6432205150096331                    1.02   
148826        6432205148853228                    4.10   
148827                     NaN                     NaN   

                            Unnamed: 6  Unnamed: 7  \
3       Серия номенклатуры.Вес металла         NaN   
4                                  NaN         NaN   
5                                47.57         NaN   
6                                15.94         NaN   
7                                 4.10         NaN   
...                                ...         ...   
148823                            1.15         NaN   
148824                            2.22         NaN   
148825                            1.02         NaN   
148826                            4.10         NaN   
148827                             NaN         NaN   

                                      Unnamed: 8  \
3       Серия номенклатуры.Вес металла в чистоте   
4                                            NaN   
5                                          44.01   
6                                          14.75   
7                                           3.79   
...                                          ...   
148823                                       NaN   
148824                                       NaN   
148825                                       NaN   
148826                                       NaN   
148827                                       NaN   

                                               Unnamed: 9  ...  \
3       Серия номенклатуры.Вес металла в чистоте ГИИСДМДК  ...   
4                                                     NaN  ...   
5                                                   44.01  ...   
6                                                   14.75  ...   
7                                                    3.79  ...   
...                                                   ...  ...   
148823                                                NaN  ...   
148824                                                NaN  ...   
148825                                                NaN  ...   
148826                                                NaN  ...   
148827                                                NaN  ...   

                                 Unnamed: 14            Unnamed: 15  \
3       Серия номенклатуры.Владелец.Артикул   Номенклатура.Артикул    
4                                        NaN                    NaN   
5                                НЦ 22-002-6            НЦ 22-002-6   
6                                НЦ 22-203-6            НЦ 22-203-6   
7                                НЦ 22-078-6            НЦ 22-078-6   
...                            

In [409]:
df_utd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148825 entries, 3 to 148827
Data columns (total 24 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   Unnamed: 0   148824 non-null  object 
 1   Unnamed: 1   0 non-null       float64
 2   Unnamed: 2   0 non-null       object 
 3   Unnamed: 3   148823 non-null  object 
 4   Unnamed: 4   148805 non-null  object 
 5   Unnamed: 5   148823 non-null  object 
 6   Unnamed: 6   148823 non-null  object 
 7   Unnamed: 7   0 non-null       float64
 8   Unnamed: 8   141129 non-null  object 
 9   Unnamed: 9   141036 non-null  object 
 10  Unnamed: 10  141092 non-null  object 
 11  Unnamed: 11  140353 non-null  object 
 12  Unnamed: 12  137369 non-null  object 
 13  Unnamed: 13  135087 non-null  object 
 14  Unnamed: 14  148360 non-null  object 
 15  Unnamed: 15  148360 non-null  object 
 16  Unnamed: 16  148823 non-null  object 
 17  Unnamed: 17  141155 non-null  object 
 18  Unnamed: 18  148823 non-

In [221]:
df_utd.drop(columns=['Unnamed: 1','Unnamed: 2','Unnamed: 7'], inplace=True)

In [222]:
df_utd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148825 entries, 3 to 148827
Data columns (total 21 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   Unnamed: 0   148824 non-null  object
 1   Unnamed: 3   148823 non-null  object
 2   Unnamed: 4   148805 non-null  object
 3   Unnamed: 5   148823 non-null  object
 4   Unnamed: 6   148823 non-null  object
 5   Unnamed: 8   141129 non-null  object
 6   Unnamed: 9   141036 non-null  object
 7   Unnamed: 10  141092 non-null  object
 8   Unnamed: 11  140353 non-null  object
 9   Unnamed: 12  137369 non-null  object
 10  Unnamed: 13  135087 non-null  object
 11  Unnamed: 14  148360 non-null  object
 12  Unnamed: 15  148360 non-null  object
 13  Unnamed: 16  148823 non-null  object
 14  Unnamed: 17  141155 non-null  object
 15  Unnamed: 18  148823 non-null  object
 16  Unnamed: 19  148734 non-null  object
 17  Unnamed: 20  148823 non-null  object
 18  Unnamed: 21  148825 non-null  object
 19  Un

In [223]:
dict_col = df_utd.iloc[0,:-3].to_dict()
dict_col

{'Unnamed: 0': 'Номенклатура.Код',
 'Unnamed: 3': 'Склад',
 'Unnamed: 4': 'Серия номенклатуры.УИН',
 'Unnamed: 5': 'Серия номенклатуры.Вес',
 'Unnamed: 6': 'Серия номенклатуры.Вес металла',
 'Unnamed: 8': 'Серия номенклатуры.Вес металла в чистоте',
 'Unnamed: 9': 'Серия номенклатуры.Вес металла в чистоте ГИИСДМДК',
 'Unnamed: 10': 'Серия номенклатуры.Вес металла ГИИС ДМДК',
 'Unnamed: 11': 'Серия номенклатуры.Характеристика номенклатуры.Полное наименование',
 'Unnamed: 12': 'Серия номенклатуры.Размер',
 'Unnamed: 13': 'Размер',
 'Unnamed: 14': 'Серия номенклатуры.Владелец.Артикул ',
 'Unnamed: 15': 'Номенклатура.Артикул ',
 'Unnamed: 16': 'Номенклатура.Текстовое описание',
 'Unnamed: 17': 'Номенклатура.Проба',
 'Unnamed: 18': 'Номенклатура.Вид номенклатуры',
 'Unnamed: 19': 'Номенклатура.Номенклатурная группа',
 'Unnamed: 20': 'Номенклатура.Наименование'}

In [224]:
dict_col['Unnamed: 21']='Количество'
dict_col['Unnamed: 22'] = 'Вес'   
dict_col['Unnamed: 23'] = 'Стоимость'

In [225]:
dict_col

{'Unnamed: 0': 'Номенклатура.Код',
 'Unnamed: 3': 'Склад',
 'Unnamed: 4': 'Серия номенклатуры.УИН',
 'Unnamed: 5': 'Серия номенклатуры.Вес',
 'Unnamed: 6': 'Серия номенклатуры.Вес металла',
 'Unnamed: 8': 'Серия номенклатуры.Вес металла в чистоте',
 'Unnamed: 9': 'Серия номенклатуры.Вес металла в чистоте ГИИСДМДК',
 'Unnamed: 10': 'Серия номенклатуры.Вес металла ГИИС ДМДК',
 'Unnamed: 11': 'Серия номенклатуры.Характеристика номенклатуры.Полное наименование',
 'Unnamed: 12': 'Серия номенклатуры.Размер',
 'Unnamed: 13': 'Размер',
 'Unnamed: 14': 'Серия номенклатуры.Владелец.Артикул ',
 'Unnamed: 15': 'Номенклатура.Артикул ',
 'Unnamed: 16': 'Номенклатура.Текстовое описание',
 'Unnamed: 17': 'Номенклатура.Проба',
 'Unnamed: 18': 'Номенклатура.Вид номенклатуры',
 'Unnamed: 19': 'Номенклатура.Номенклатурная группа',
 'Unnamed: 20': 'Номенклатура.Наименование',
 'Unnamed: 21': 'Количество',
 'Unnamed: 22': 'Вес',
 'Unnamed: 23': 'Стоимость'}

In [226]:
df_utd.rename(columns=dict_col, inplace=True)
df_utd

Номенклатура.Код          Склад  Серия номенклатуры.УИН  \
3       Номенклатура.Код          Склад  Серия номенклатуры.УИН   
4                    NaN            NaN                     NaN   
5              000005200   Благовещенск        6432205037320158   
6              000005201   Благовещенск        6432205037445837   
7              000005202   Благовещенск        6432205037404896   
...                  ...            ...                     ...   
148823         000186413  УР Бриллианты        6432205148666206   
148824         000186414  УР Бриллианты        6432205150175616   
148825         000186415  УР Бриллианты        6432205150096331   
148826         000186416  УР Бриллианты        6432205148853228   
148827             Итого            NaN                     NaN   

        Серия номенклатуры.Вес  Серия номенклатуры.Вес металла  \
3       Серия номенклатуры.Вес  Серия номенклатуры.Вес металла   
4                          NaN                             NaN   
5                        47.57                           47.57   
6                        15.94                           15.94   
7                          4.1                             4.1   
...                        ...                             ...   
148823                    1.15                            1.15   
148824                    2.22                            2.22   
148825                    1.02                            1.02   
148826                     4.1                             4.1   
148827                     NaN                             NaN   

        Серия номенклатуры.Вес металла в чистоте  \
3       Серия номенклатуры.Вес металла в чистоте   
4                                            NaN   
5                                         44.007   
6                                         14.746   
7                                          3.793   
...                                          ...   
148823                                       NaN   
148824                                       NaN   
148825                                       NaN   
148826                                       NaN   
148827                                       NaN   

        Серия номенклатуры.Вес металла в чистоте ГИИСДМДК  \
3       Серия номенклатуры.Вес металла в чистоте ГИИСДМДК   
4                                                     NaN   
5                                                  44.007   
6                                                  14.746   
7                                                   3.793   
...                                                   ...   
148823                                                NaN   
148824                                                NaN   
148825                                                NaN   
148826                                                NaN   
148827                                                NaN   

        Серия номенклатуры.Вес металла ГИИС ДМДК  \
3       Серия номенклатуры.Вес металла ГИИС ДМДК   
4                                            NaN   
5                                          47.57   
6                                          15.94   
7                                            4.1   
...                                          ...   
148823                                       NaN   
148824                                       NaN   
148825                                       NaN   
148826                                       NaN   
148827                                       NaN   

       Серия номенклатуры.Характеристика номенклатуры.Полное наименование  \
3       Серия номенклатуры.Характеристика номенклатуры...                   
4                                                     NaN                   
5                                              Без камней                   
6                                              Без камней                   
7                                         

In [227]:
df_utd.drop([3,4], inplace=True)


In [228]:
df_utd.iloc

In [229]:
df_utd = df_utd.iloc[:-1].copy(deep=True)

In [230]:
len(df_utd['Номенклатура.Код'].unique()), len(df_utd)

(148822, 148822)

In [231]:
df_utd.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 148822 entries, 5 to 148826
Data columns (total 21 columns):
 #   Column                                                              Non-Null Count   Dtype 
---  ------                                                              --------------   ----- 
 0   Номенклатура.Код                                                    148822 non-null  object
 1   Склад                                                               148822 non-null  object
 2   Серия номенклатуры.УИН                                              148804 non-null  object
 3   Серия номенклатуры.Вес                                              148822 non-null  object
 4   Серия номенклатуры.Вес металла                                      148822 non-null  object
 5   Серия номенклатуры.Вес металла в чистоте                            141128 non-null  object
 6   Серия номенклатуры.Вес металла в чистоте ГИИСДМДК                   141035 non-null  object
 7   Серия номен

In [233]:
df_utd.loc[df_utd['Серия номенклатуры.Характеристика номенклатуры.Полное наименование'].isna(),
           ['Серия номенклатуры.Характеристика номенклатуры.Полное наименование']] = ''

## подсчет длина вставки == 100

In [234]:
df_utd['Серия номенклатуры.Характеристика номенклатуры.Полное наименование'].apply(lambda x: True if len(x) == 100 else False).sum()

12

In [235]:
df_utd.loc[df_utd['Серия номенклатуры.Характеристика номенклатуры.Полное наименование']\
 .apply(lambda x: True if len(x) == 100 else False),
  'Серия номенклатуры.Характеристика номенклатуры.Полное наименование'].unique()

array(['Хромдиопсид Гранат ТопазГ Цитрин 4Кр57-0,079-6/6А 4Кр57-0,063-6/6А 4Кр57-0,043-4/5А 3Кр57-0,153-3/5А',
       '1Кр57-0,860-5/8А 1Кр57-0,030-3/5Б 1Кр57-0,030-3/5Б 1Кр57-0,030-3/3Б 3Кр57-0,010-3/5Б 3Кр57-0,090-3/6',
       '1КР57-0,810-4/8А 2КР57-0,070-4/5А 1КР57-0,030-4/5А 2КР57-0,060-4/6А 2КР57-0,060-3/6Б 3КР57-0,090-4/3',
       '1КР17-0,003-3/3А 8КР57-0,029-4/6А 5КР57-0,015-4/6А 5КР57-0,017-4/6А 4КР57-0,016-6/6Б 9КР57-0,041-6/6',
       '17Кр57-0,2-1/4А 99Кр57-1,78-1/7А 65Кр57-0,89-3/6А 4Кр57-0,03-3/6А 15Кр57-0,16-7/7Б 165Кр57-2,82-7/7Б',
       '1Г56-1,15-5/4Б 1Ов57-0,51-8-5/7а 1Ов57-0,-8.5/4А 10кр57-0,14-3/5А 36кр57-0,35-3/5А 147кр57-0,83-3/5А',
       'рубинВ 9кр57-0,8-4/5А 2кр57-0,03-4/5А 2кр57-0,04-4/5А 2кр57-0,05-4/5А 2кр57-0,08-4/5А 4кр57-0,19-4/5',
       'ЖемчугБ 10КР57-0,180-3/5А 11КР57-0,470-2/2А 9КР57-0,320-2/2А 10КР57-0,290-2/2А 6КР57-0,180-2/3А 38КР',
       '1Кр57-0,080-3/5А 1Кр57-0,090-2/5А 1Кр57-0,070-2/5А 6Кр57-0,54-2/5А 3Кр57-0,020-3/4А 5Кр57-0,046-2/4А',
 

In [236]:
len(_)

10

In [238]:
df_utd['проблема вставок ЮТД 100'] = False
df_utd.loc[df_utd['Серия номенклатуры.Характеристика номенклатуры.Полное наименование'].apply(lambda x: True if len(x) == 100 else False),
          'проблема вставок ЮТД 100'] = True

In [239]:
df_utd['проблема вставок ЮТД 100'].sum()

12

## вывод кода ЦБ

In [240]:
df_utd['Код'] = df_utd['Номенклатура.Текстовое описание'].apply(lambda x: x.split(';')[0])

In [241]:
len(df_utd['Код'].unique())

148802

In [242]:
df_utd = df_utd[~df_utd['Код'].duplicated(keep='first')].copy(deep=True)#.sort_values('Номенклатура.Текстовое описание')

In [243]:
#_.to_excel('файл для Василия_дупликаты 20 изделий в ЮТД.xlsx')

# merge. У ЮТД переименовываем склад на склад_ютд

In [310]:
df_1c_res

Код Старый код (Справочник "Номенклатура" (Общие))  \
0      ЦБ-00200503                                            NaN   
1      ЦБ-00002854                                     МС00057368   
2      ЦБ-00019951                                     МС00152100   
3      ЦБ-00019953                                     МС00146311   
4      ЦБ-00111074                                     МС00111422   
...            ...                                            ...   
34329  ЦБ-00182016                                     0000032928   
34330  ЦБ-00182017                                     0000032929   
34331  ЦБ-00182019                                     0000032934   
34332  ЦБ-00182020                                     0000032935   
34333  ЦБ-00182021                                     0000032936   

                   Склад Склад.Склад / Магазин  \
0           Благовещенск     Розничный магазин   
1           Благовещенск     Розничный магазин   
2           Благовещенск     Розничный магазин   
3           Благовещенск     Розничный магазин   
4           Благовещенск     Розничный магазин   
...                  ...                   ...   
34329  Ювелирные изделия         Оптовый склад   
34330  Ювелирные изделия         Оптовый склад   
34331  Ювелирные изделия         Оптовый склад   
34332  Ювелирные изделия         Оптовый склад   
34333  Ювелирные изделия         Оптовый склад   

             Вставки (Справочник "Номенклатура" (Общие)) Вид номенклатуры  \
0                                      45КР57-0,520-4/5А     Т ДМ Юи с ДК   
1      2Кр57-0,083-2/2А 4Кр57-0,084-2/2А 6Кр57-0,280-...     Т ДМ Юи с ДК   
2                                       1Кр57-0,086-4/5А     П ДМ Юи с ДК   
3                                        3Кр57-0,06-4/5А     П ДМ Юи с ДК   
4      1Кр57-0,050-4/5А 1Кр57-0,031-6/6А 1Кр57-0,014-...     Т ДМ Юи с ДК   
...                                                  ...              ...   
34329                                                NaN     Т ДМ Юи с ДК   
34330                                                NaN     Т ДМ Юи с ДК   
34331                                                NaN     Т ДМ Юи с ДК   
34332                                                NaN     Т ДМ Юи с ДК   
34333                                                NaN     Т ДМ Юи с ДК   

      Тип номенклатуры Номенклатура                           Группа  \
0                Товар      Браслет  спецификация №1 от 27.01.2022г.   
1                Товар      Браслет                         Товар ДК   
2                Товар        Брошь                            Броши   
3                Товар        Брошь                          накл.№5   
4                Товар        Брошь                            51 ДК   
...                ...          ...                              ...   
34329            Товар  серьги с ДК        Акт №1 АК "Золото Якутии"   
34330            Товар  серьги с ДК        Акт №1 АК "Золото Якутии"   
34331            Товар  серьги с ДК        Акт №1 АК "Золото Якутии"   
34332            Товар  серьги с ДК        Акт №1 АК "Золото Якутии"   
34333            Товар  серьги с ДК        Акт №1 АК "Золото Якутии"   

      Заводской номер (Справочник "Номенклатура" (Общие))  ...  \
0                                                    512   ...   
1                                               30575,51   ...   
2                                                1191890   ...   
3                                                   1094   ...   
4                                                    422   ...   
...                                                  ...   ...   
34329                                           01-12308   ...   
34330                                           01-12309   ...   
34331                                           01-14045   ...   
34332                                           01-14046   ...   
34333                                           01-14047   ...   

  

In [311]:
df_utd.rename(columns={'Склад':'Склад_ютд'}, inplace=True)

In [312]:
#Присоединяем к df_1c_res
df_1c_res = pd.merge(df_1c_res,
         df_utd,
         how='left',
         right_on=['Код'],
         left_on=['Код']
         )

In [313]:
df_1c_res.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 34334 entries, 0 to 34333
Data columns (total 63 columns):
 #   Column                                                                   Non-Null Count  Dtype  
---  ------                                                                   --------------  -----  
 0   Код                                                                      34334 non-null  object 
 1   Старый код (Справочник "Номенклатура" (Общие))                           28294 non-null  object 
 2   Склад                                                                    34334 non-null  object 
 3   Склад.Склад / Магазин                                                    34334 non-null  object 
 4   Вставки (Справочник "Номенклатура" (Общие))                              24346 non-null  object 
 5   Вид номенклатуры                                                         34334 non-null  object 
 6   Тип номенклатуры                                                      

In [324]:
# Потом проверить
df_1c_res[df_1c_res['Номенклатура.Код'].isna()]

Код Старый код (Справочник "Номенклатура" (Общие))  \
951    ЦБ-00201416                                            NaN   
4280   ЦБ-00201418                                            NaN   
4989   ЦБ-00201422                                            NaN   
6668   ЦБ-00201420                                            NaN   
7933   ЦБ-00201421                                            NaN   
...            ...                                            ...   
15649  ЦБ-00201415                                            NaN   
16851  ЦБ-00201419                                            NaN   
18372  ЦБ-00075259                                     МС00019549   
26684  ЦБ-00219596                                            NaN   
34182  ЦБ-00201586                                            NaN   

                   Склад Склад.Склад / Магазин  \
951         Благовещенск     Розничный магазин   
4280             Иркутск     Розничный магазин   
4989             Кружало     Розничный магазин   
6668          Ленина, 28     Розничный магазин   
7933            Ленина23     Розничный магазин   
...                  ...                   ...   
15649          Хабаровск     Розничный магазин   
16851        Центральный     Розничный магазин   
18372  Ювелирные изделия         Оптовый склад   
26684  Ювелирные изделия         Оптовый склад   
34182  Ювелирные изделия         Оптовый склад   

             Вставки (Справочник "Номенклатура" (Общие)) Вид номенклатуры  \
951    36КР57-0,137-4/5А 24КР57-0,173-4/5А 4КР57-0,12...     Т ДМ Юи с ДК   
4280   36КР57-0,137-4/5А 24КР57-0,173-4/5А 4КР57-0,12...     Т ДМ Юи с ДК   
4989   18КР57-0,070-4/5А 12КР57-0,090-4/5А 2КР57-0,06...     Т ДМ Юи с ДК   
6668   18КР57-0,070-4/5А 12КР57-0,090-4/5А 2КР57-0,06...     Т ДМ Юи с ДК   
7933   18КР57-0,070-4/5А 12КР57-0,090-4/5А 2КР57-0,06...     Т ДМ Юи с ДК   
...                                                  ...              ...   
15649                  6КР57-0,197-4/5А 1КР57-0,050-4/5А     Т ДМ Юи с ДК   
16851  18КР57-0,070-4/5А 12КР57-0,090-4/5А 2КР57-0,06...     Т ДМ Юи с ДК   
18372                                                NaN     Т ДМ Юи с ДК   
26684  1МодГ-1,010-4/5А 36Кр57-0,425-5/5А 8Кр57-0,260...     П ДМ Юи с ДК   
34182                                   2КР57-0,050-4/5А     Т ДМ Юи с ДК   

      Тип номенклатуры Номенклатура                           Группа  \
951              Товар       Серьги  спецификация №1 от 27.01.2022г.   
4280             Товар       Серьги  спецификация №1 от 27.01.2022г.   
4989             Товар       Кольцо  спецификация №1 от 27.01.2022г.   
6668             Товар       Кольцо  спецификация №1 от 27.01.2022г.   
7933             Товар       Кольцо  спецификация №1 от 27.01.2022г.   
...                ...          ...                              ...   
15649            Товар     Подвеска  спецификация №1 от 27.01.2022г.   
16851            Товар       Кольцо  спецификация №1 от 27.01.2022г.   
18372            Товар       Кольцо                           126 дк   
26684            Товар       Кольцо                        эксклюзив   
34182            Товар       Серьги  спецификация №1 от 27.01.2022г.   

      Заводской номер (Справочник "Номенклатура" (Общие))  ...  \
951                                                 1427   ...   
4280                                                1429   ...   
4989                                                1433   ...   
6668                                                1431   ...   
7933                                                1432   ...   
...                                                  ...   ...   
15649                                               1426   ...   
16851                                               1430   ...   
18372                                            1177-12   ...   
26684                                   3562200005684812   ...   
34182                                                674   ...   

  

In [204]:
# Запись для сверки врем файл
df_1c_res.to_excel(folder_res+'сравнение ЮТД и УТ по ДК.xlsx')

In [333]:
folder_res

'../for_script08-10-2023/'

In [326]:
#  имеется ошибки вставок в ЮТД в виде слова 'ЛОЖЬ'
df_utd['Серия номенклатуры.Характеристика номенклатуры.Полное наименование'] == 'ЛОЖЬ'

5         False
6         False
7         False
8         False
9         False
          ...  
148822    False
148823    False
148824    False
148825    False
148826    False
Name: Серия номенклатуры.Характеристика номенклатуры.Полное наименование, Length: 148802, dtype: bool

In [327]:
sum(_)

23

In [329]:
df_1c_res['Штрихкод'].isna().sum()

222

In [331]:
df_1c_res['Вставки010123'].isna().sum()

0

# Запись результата в файл 2

In [332]:
folder_res = '../for_script08-10-2023/'
folder_res

'../for_script08-10-2023/'

In [84]:
#запись промрезультата для проверки на excel
df_1c_res.to_excel(folder_res+'Остатки1С__все_склады_ДК_2023-10-08_10-12.xlsx')

In [85]:
#df_1c_res.loc[df_1c_res['Код'] =='ЦБ-00040270','Сводка Вставки' ].to_list()

# del

In [ ]:
  ['Т ДМ Юи с ДК', 'Т ДМ Юи сер. с ДК','П ДМ Юи с ДК','Т ДМ Юи Каст ДК','П ДМ Юи Каст ДК']

In [81]:
(len(df_vst2.loc[(df_vst2['Вид ном'].isin( ['Т ДМ Юи с ДК', 'Т ДМ Юи сер. с ДК','П ДМ Юи с ДК',
                                                   'Т ДМ Юи Каст ДК','П ДМ Юи Каст ДК'])),'Код'].unique()),
                 len(df_vst2.loc[(df_vst2['Вид ном'].isin(  ['Т ДМ Юи с ДК', 'Т ДМ Юи сер. с ДК','П ДМ Юи с ДК',
                                                             'Т ДМ Юи Каст ДК','П ДМ Юи Каст ДК'])),'Код']))

(36398, 36398)

In [84]:
df_vst2_DK = df_vst2.loc[(df_vst2['Вид ном'].isin(['Т ДМ Юи с ДК', 'Т ДМ Юи сер. с ДК','П ДМ Юи с ДК',
                                                             'Т ДМ Юи Каст ДК','П ДМ Юи Каст ДК']))]

In [88]:
df_vst2_DK.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 36398 entries, 1758 to 153277
Data columns (total 33 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Склад                      36398 non-null  object 
 1   Штрихкод                   36388 non-null  float64
 2   Старый код                 30683 non-null  object 
 3   Код                        36398 non-null  object 
 4   Наименование               36398 non-null  object 
 5   Артикул                    36397 non-null  object 
 6   Зав Ном                    35861 non-null  object 
 7   Вид ном                    36398 non-null  object 
 8   Проба                      36398 non-null  float64
 9   Цвет металла               36304 non-null  object 
 10  Кол                        36398 non-null  float64
 11  Вес                        36398 non-null  float64
 12  Закуп                      36398 non-null  float64
 13  Розн                       21664 non-null 

In [40]:
len(df_1c_res['Код'].unique()), len(df_1c_res)

(34528, 34528)

In [87]:
len(set(df_1c_res['Код']) - set(df_vst2_DK['Код']))

214

In [89]:
36398-34528

1870

In [ ]:
'Родитель1', 'Родитель2', 'Родитель3', 'Родитель4', 'Родитель5'

o	Данные по вставкам надежнее чем с 1с (на 1 янв 2023г) 1. Вводные\Данные по вставкам\23-01-01 Сводка ЮИ.xlsx

In [ ]:
files = os.listdir(folder+'1. Вводные/Данные по вставкам/')
files

[x for x in enumerate(files)]

df_vst = pd.read_excel(folder+'1. Вводные/Данные по вставкам/'+files[0])#,header=9)#, dtype={5: str}

df_vst.columns

len(df_vst['Код'].unique()), len(df_vst)

len(df_1c_res['Код'].unique()), len(df_1c_res)

#Присоединяем к df_1c
df_1c_res = pd.merge(df_1c_res,
         df_vst[['Код','Вставки', 'Штрихкод', 'Старый код',
                 'Родитель1', 'Родитель2', 'Родитель3',
                 'Родитель4', 'Родитель5' ]],
         how='left',
         right_on=['Код'],
         left_on=['Код']
         )

df_1c_res['Штрихкод'].isna().sum()